src/hotspot/share/gc/g1/g1FullGCPrepareTask.inline.hpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_G1_G1FULLGCPREPARETASK_INLINE_HPP
+#define SHARE_GC_G1_G1FULLGCPREPARETASK_INLINE_HPP
+
+#include "gc/g1/g1FullGCPrepareTask.hpp"
+
+#include "gc/g1/g1CollectedHeap.inline.hpp"
+#include "gc/g1/g1FullCollector.hpp"
+#include "gc/g1/g1FullGCCompactionPoint.hpp"
+#include "gc/g1/g1FullGCScope.hpp"
+#include "gc/g1/heapRegion.inline.hpp"
+
+template<bool is_humongous>
+void G1DetermineCompactionQueueClosure::free_pinned_region(HeapRegion* hr) {
+  _found_empty_regions = true;
+  if (is_humongous) {
+    _g1h->free_humongous_region(hr, nullptr);
+  } else {
+    _g1h->free_region(hr, nullptr);
+  }
+  _collector->set_free(hr->hrm_index());
+  add_to_compaction_queue(next_compaction_point(), hr);
+}
+
+inline bool G1DetermineCompactionQueueClosure::should_compact(HeapRegion* hr) const {
+  // There is no need to iterate and forward objects in pinned regions ie.
+  // prepare them for compaction.
+  if (hr->is_pinned()) {
+    return false;
+  }
+  size_t live_words = _collector->live_words(hr->hrm_index());
+  size_t live_words_threshold = _collector->scope()->region_compaction_threshold();
+  // High live ratio region will not be compacted.
+  return live_words <= live_words_threshold;
+}
+
+inline uint G1DetermineCompactionQueueClosure::next_worker() {
+  uint result = _cur_worker;
+  _cur_worker = (_cur_worker + 1) % _collector->workers();
+  return result;
+}
+
+inline G1FullGCCompactionPoint* G1DetermineCompactionQueueClosure::next_compaction_point() {
+  return _collector->compaction_point(next_worker());
+}
```

Comment:
I don't see much benefit of having such separated in two methods; this is the sole call site of `next_worker()` after all. Instead of `%`, manually setting it to zero on reaching `workers()` is fairly readable as well.

Reply:
This is a fairly trivial implementation for `next_worker()` to fix the issue at hand. As suggested by the description, there needs to be some more thought put into it to handle e.g. imbalance of region occupancies or humongous regions.
Unless you insist I would like to keep the current implementation as is.

Reply:
> I would like to keep the current implementation as is.

Sure.